include/h2o/memory.h

```diff
@@ -301,6 +305,13 @@ void h2o_append_to_null_terminated_list(void ***list, void *element);
 
 /* inline defs */
 
+inline void *h2o_memcpy(void *dst, const void *src, size_t n)
+{
+    if (src)
```

Comment:
> Generally speaking, I do not think that we need to avoid UB when we have a logic flaw  

I completely agree. This PR aims at avoiding possibly dangerous compiler optimizations

Here's an example showcasing what I mean: https://godbolt.org/g/1Fb2fg : the `if` check in `f` has been completely removed by the compiler because it assumes that it's not possible that `src` becomes `NULL`, since it's passed to a `f` which tags `src` as `nonnull`. 

IOW:

```
$ cat main.c                                                                                                                                                                                                                                                                                                                #include <stdio.h>                                                                                                                                                                                                                                                                                                            #include <stddef.h>                                                                                                                                                                                                                                                                                                           #include <stdlib.h>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         void *f(void *dst, void *src, size_t n) __attribute__((nonnull(1, 2)));                                                                                                                                                                                                                                                       int main(void)                                                                                                                                                                                                                                                                                                                {                                                                                                                                                                                                                                                                                                                                 
    char buf[64];                                                                                                                                                                                                                                                                                                                 
    fprintf(stderr, "%p\n", f(buf, getenv("THIS_IS_NOT_SET"), 0));                                                                                                                                                                                                                                                                           
    return 0;                                                                                                                                                                                                                                                                                                                 }
$ cat lib.c                                                                                                                                                                                                                                                                                                                 #include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>

void g(void)                                                                                                                                                                                                                                                                                                                  {                                                                                                                                                                                                                                                                                                                                
     printf("called\n");                                                                                                                                                                                                                                                                                                        }                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           void *f(void *dst, void *src, size_t n) __attribute__((nonnull(1, 2)));                                                                                                                                                                                                                                                       void *f(void *dst, void *src, size_t n)                                                                                                                                                                                                                                                                                       {                                                                                                                                                                                                                                                                                                                                 
     if (src == NULL)                                                                                                                                                                                                                                                                                                                  
         g();                                                                                                                                                                                                                                                                                                                      
    memcpy(dst, src, n);                                                                                                                                                                                                                                                                                                      }
$ gcc -O1 main.c lib.c ; ./a.out
called
0x7fff2f54c590
$ gcc -O2 main.c lib.c ; ./a.out
0x7ffe796cf510

```